### ISSUE TITLE
Bug: Negated patterns are considered explicit match

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#19813]
### Environment

Node version: v22.14.0
npm version: v10.9.2
Local ESLint version: v9.28.0
Global ESLint version: no
Operating System: windows

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
export default [
    {
        files: ["!foo.js"],
        rules: {
            "no-undef": 2
        }
    }
];
```
</details>

```js
// foo.js
foo
```

### What did you expect to happen?

I'd expect only .js, .mjs, and .cjs files to be linted, and the config object to apply to all files except `foo.js`.

### What actually happened?

The `"!foo.js"` pattern is considered an explicit match, so it enables linting of all files.

```
$ npx eslint .

C:\projects\tmp\tmp\package-lock.json
  2:9  error  Parsing error: Unexpected token :

C:\projects\tmp\tmp\package.json
  2:9  error  Parsing error: Unexpected token :

✖ 2 problems (2 errors, 0 warnings)
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-skkgpzj7

### Participation

- [x] I am willing to submit a pull request for this issue.

**Comment by lumirlumir** [comment]
It's weird. I would like to add some additional contexts.

When I set the files to `["*/.json"]`, the config inspector shows the correct files.

![Image](https://github.com/user-attachments/assets/537fabba-c8c5-4d59-bf41-367088273c42)

Also, when I set the files to `["foo.js"]`, the config inspector shows the correct files.

![Image](https://github.com/user-attachments/assets/7bbec3dd-43b1-4eca-8e90-b2ebdcc27186)

However, when I set it to `["!foo.js"]`, no files match, but ESLint report errors on `package.json`, `package-lock.json` and `.gitignore`.

![Image](https://github.com/user-attachments/assets/a9cc5d94-ddba-4ed3-919b-3fe7c14f5cf4)

![Image](https://github.com/user-attachments/assets/f2fbccf2-9c71-4b8a-a4b8-07f8b58b5e77)

**Comment by nzakas** [comment]
To clarify what's being reported from a user perspective ("explicit match" is a term used internally in the code and isn't a user-facing term): The design of negated patterns is meant to be exclusionary of the files being linted. So if you're using the default file patterns from ESLint, `!foo.js` means "lint all files as usual and then omit `foo.js` from that list."

What's actually happening is ESLint is ignoring the default file patterns and instead linting every file that doesn't match the pattern `foo.js`, leading to linting of files that shouldn't be linted.

I agree, this is a bug.

**Comment by mdjermanovic** [comment]
I'll submit a PR for this when https://github.com/eslint/rewrite/pull/218 is merged.
